private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/maven/MavenResolver.java

```diff
@@ -0,0 +1,423 @@
+        .forEach(toReturn::add);
+
+    return toReturn.build();
+  }
+
+  private List<Dependency> resolveArtifactsFromBoms(
+      RepositorySystem system,
+      RepositorySystemSession session,
+      List<RemoteRepository> repositories,
+      List<Dependency> amendedBoms) {
+    Set<Dependency> managedDependencies = new HashSet<>();
+
+    for (Dependency bom : amendedBoms) {
+      ArtifactDescriptorRequest request =
+          new ArtifactDescriptorRequest(bom.getArtifact(), repositories, JavaScopes.COMPILE);
+      try {
+        ArtifactDescriptorResult result = system.readArtifactDescriptor(session, request);
+        managedDependencies.addAll(result.getManagedDependencies());
+      } catch (ArtifactDescriptorException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    return ImmutableList.copyOf(managedDependencies);
+  }
+
+  private List<Dependency> addGlobalExclusions(
+      Set<Exclusion> globalExclusions, List<Dependency> dependencies) {
+    return dependencies.stream()
+        .map(
+            dep -> {
+              Set<Exclusion> allExclusions = new HashSet<>(globalExclusions);
+              allExclusions.addAll(dep.getExclusions());
+
+              return new Dependency(
+                  dep.getArtifact(), dep.getScope(), dep.getOptional(), allExclusions);
+            })
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  private Graph<Coordinates> buildGraph(
+      Map<Coordinates, Coordinates> remappings, Collection<DependencyNode> directDependencies) {
+    MutableGraph<Coordinates> toReturn = GraphBuilder.directed().allowsSelfLoops(true).build();
+    DependencyVisitor collector =
+        new TreeDependencyVisitor(
+            new DependencyNodeVisitor(
+                node -> {
+                  final DependencyNode actualNode = getDependencyNode(node);
+
+                  Artifact artifact = amendArtifact(actualNode.getArtifact());
+                  Coordinates from = MavenCoordinates.asCoordinates(artifact);
+                  Coordinates remapped = remappings.getOrDefault(from, from);
+                  toReturn.addNode(remapped);
+
+                  actualNode.getChildren().stream()
+                      .map(DependencyNode::getArtifact)
+                      .map(this::amendArtifact)
+                      .map(MavenCoordinates::asCoordinates)
+                      .map(c -> remappings.getOrDefault(c, c))
+                      .forEach(
+                          to -> {
+                            toReturn.addNode(to);
+                            toReturn.putEdge(remapped, to);
+                          });
+                }));
+    directDependencies.forEach(node -> node.accept(collector));
+
+    return ImmutableGraph.copyOf(toReturn);
+  }
+
+  private static DependencyNode getDependencyNode(DependencyNode node) {
+    Map<?, ?> data = node.getData();
+    if (data != null) {
+      Object winner = data.get(ConflictResolver.NODE_DATA_WINNER);
```

Comment:
Can you add a comment about what is going on here?  Feels like maybe this function could have a more descriptive name for what it is doing.

Reply:
Done